### DISCUSSION TITLE
xunit >2.7.0 threading changes causing test to fail

### DISCUSSION THREAD

**Discussion by martincostello** [xunit/xunit#3024] [Question] [answered]
We've had a dependabot PR open to update Polly from xunit 2.7.0 open for a while now: https://github.com/App-vNext/Polly/pull/2186

We've been unable to merge this PR as it is causing one of our tests to fail - specifically, [this test](https://github.com/App-vNext/Polly/blob/c6aaafb5f8ff7a91ae55cf95ecd12a8fe65f6b40/test/Polly.Core.Tests/Hedging/HedgingStrategyOptionsTests.cs#L22).

We have a test that aims to ensure that if a hedged action is performed synchronously that it is run on a new thread compared to the caller, but that test is failing as that's no longer happening.

Looking at [the diff](https://github.com/xunit/xunit/compare/2.7.0...2.9.0) I saw some changes related to async and thread pool sizes, but I haven't been able to reconfigure xunit to behave the same as it did in 2.7.0.

Are we doing something wrong in Polly and this test should never have reliably worked, or is there a setting I'm missing to restore the behaviour, or is this just a bug in xunit?

**Comment by bradwilson** [comment, marked as answer]
You're running afoul of the new parallelism algorithm change.

The difference is probably [here](https://github.com/xunit/xunit/blob/416ed8182d9b6a4a712563e6d7519a7c8281588e/src/xunit.execution/Sdk/Frameworks/Runners/XunitTestAssemblyRunner.cs#L230-L233):

* The `aggressive` algorithm (the one that was the default--and only choice--in < 2.8.0) uses a [custom thread pool](https://github.com/xunit/xunit/blob/416ed8182d9b6a4a712563e6d7519a7c8281588e/src/xunit.execution/Sdk/Frameworks/Runners/XunitTestAssemblyRunner.cs#L85-L92) (via [`MaxConcurrencySyncContext`](https://github.com/xunit/xunit/blob/416ed8182d9b6a4a712563e6d7519a7c8281588e/src/xunit.execution/Sdk/MaxConcurrencySyncContext.cs#L28-L30)) to run tests from, which is limited in size to the "max parallel threads" setting. This thread pool is unrelated to the system thread pool, and more importantly, the only work that can ever get scheduled on these threads is unit tests.

* The `conservative` algorithm (the default >= 2.8.0) [uses the system thread pool](https://github.com/xunit/xunit/blob/416ed8182d9b6a4a712563e6d7519a7c8281588e/src/xunit.execution/Sdk/Frameworks/Runners/XunitTestAssemblyRunner.cs#L299-L336) to run tests from, and limits the tests by using a semaphore. Additionally, we ensure that the system thread pool has a minimum thread count of 4, since it's a shared resource and CI machines will sometimes only have 1 or 2 CPU cores.

So an obvious way I could see this test failing is: if your test is now running on the system thread pool because of the default `conservative` algorithm, and then you schedule work to be run on the system thread pool, when you hit `await` in the test method, that thread goes back into the pool to be used for whatever needs it, and it's possible that your scheduled background work ends up using that same thread to do the work, because the test hasn't blocked it. (If this is the case, then the likelihood of this goes up when (a) you have fewer threads to use, and/or (b) all the other threads are currently in use.) This would never happen with the `aggressive` algorithm because the thread that got given up by the test isn't part of the system thread pool.

Assuming everything I've said here is what's happening (you sometimes have to make guesses when it comes to threading stuff 😂), then yes, you wrote this test (unintentionally) based on an implementation detail that stopped being true by default in 2.8.0.

The good news is I was able to (mostly) get the failure to repro locally. After [configuring for the aggressive algorithm](https://xunit.net/docs/configuration-files#parallelAlgorithm) to revert to the old behavior, I was unable to get it fail again. So I _believe_ that my guess is correct as to the root cause here.

Which leaves you (at least) three choices:

1. Use a configuration file to revert to the `aggressive` algorithm;
2. Don't use `await` on the incomplete task and instead use a waiting mechanism that doesn't give the thread back to the system thread pool;
3. Find another way to determine that the background work got scheduled to run on the system thread pool, without comparing the thread IDs.

Let's talk about option 2 for a moment.

We have [an analyzer rule](https://xunit.net/xunit.analyzers/rules/xUnit2021) that you would have to disable in this context if you chose to use one of the built-in mechanisms, like `.GetAwaiter().GetResult()`. The reason behind this rule is to prevent thread deadlock situations, especially when using async APIs that might need to return back to the original thread. Since the change in 2.8.0, as mentioned above we have ensured that the minimum system thread pool size is 4, which means a _single test_ that took this strategy couldn't realistically deadlock forever, unless somehow three _other_ things were deadlocked as a result.

This isn't a particularly ideal situation, but it would certainly make sense with the test as written, since the test as written makes the assumption that it's _impossible_ for the background work to run on the same thread as the test. While that is true with the `aggressive` algorithm, that's no longer true with `await` and the `conservative` algorithm.

If you choose option 2, you could write the wait for the task completion in terms of Thread.Sleep or Thread.SpinWait so you don't rely on a construct that will trigger xUnit2021. You could replace:

```csharp
(await action()).Result.Should().Be(99);
```

with something like:

```csharp
var task = action();
while (!task.IsCompleted)
{
    Thread.SpinWait(5);
}

(await task).Result.Should().Be(99);
```

(I have verified that this exact code does indeed also fix the issue, without needing to reconfigure the parallelism algorithm, but with the risk that it is doing deadlock-y behavior.)

I would clearly document in the test why the code is written this way, because it would be easy for someone who doesn't know to be puzzled why a simple `await` was not sufficient. It would also make it clear that this should not become a general strategy for _lots_ of tests; it's fairly safe to do with one test, but if you used this enough across the test code base you could absolutely end up deadlocking yourself if you aren't careful (again, a situation that becomes much more likely on a typical "free" CI environment like Actions where you have just a few CPU cores, and thus a small thread pool size).

-----

Personally, in your situation, I would choose option 1 now, with an eye towards seeing if option 3 is reasonable. This has been how xUnit.net has been running since we first added support for parallelization and Tasks. You are _potentially_ giving up some value, so reviewing [the documentation](https://xunit.net/docs/running-tests-in-parallel#algorithms) would make sense when trying to choose whether to do option 1 or not, but it's by far the easiest and safest choice right now. The best long-term choice would be option 3, but without digging further into your code I honestly have no idea what a good alternative mechanism would be. Sometimes you end up bending the production code in the name of testability/observability, which is something I'll almost always choose to do personally, but lots of people become a little uncomfortable letting testability dictate the design of their code.

Let me know if you have any other questions.

**Reply by martincostello** [reply]
Thanks for the comprehensive reply Brad!

I'll look into this properly tomorrow and pick whichever option makes the most sense and let you know how I get on.

**Reply by bradwilson** [reply]
Ooh, also, I should note: the SpinWait sample I gave you would run indefinitely if for some reason the work _didn't_ get scheduled as background work and was instead waiting for this exact thread to continue. This is really the canonical case of the deadlock situation we are trying to help people avoid.

If you opt for option 2 with my sample code, you should set a limit on how long you're willing to wait before failing the test, to prevent it for waiting forever if you happen to have a bug in the scheduling algorithm.

**Reply by martincostello** [reply]
Thanks again - I've gone with option 1 for now, and that's unblocked the PR to update xunit to the latest version.

I've opened https://github.com/App-vNext/Polly/issues/2288 to track adopting option 3.

**Reply by martincostello** [reply]
A Polly contributor has kindly proposed a PR to resolve that issue. Would you mind taking a look @bradwilson? https://github.com/App-vNext/Polly/pull/2313.

It seems to be a variant of option 2 (rather than an attempt at 3).
